### ISSUE TITLE
`strat --restrict` does not work properly on fetched strata when using zsh

### ISSUE THREAD

**Issue by cptpcrd** [bedrocklinux/bedrocklinux-userland#108]
*TL;DR*: `. /bedrock/run/zprofile` being automatically added to `/etc/zsh/zshenv` breaks `strat -r zsh [...]`.

Setup: Hijacked Arch install with Void and Ubuntu fetched (problem and solution reproducible in both Void and Ubuntu); `zsh` installed and set as current user's shell
```sh
$ strat -r ubuntu bash -c 'echo $PATH' #Correctly set
/bedrock/bin:/usr/local/sbin:/usr/local/bin:/opt/sbin:/opt/bin:/usr/sbin:/usr/bin:/sbin:/bin
$ strat -r ubuntu zsh -c 'echo $PATH' #Incorrectly set
/bedrock/cross/pin/bin:/bedrock/bin:/usr/local/sbin:/usr/local/bin:/opt/sbin:/opt/bin:/usr/sbin:/usr/bin:/sbin:/bin:/bedrock/cross/bin
$ strat ubuntu cat /etc/zsh/zshenv
# /etc/zsh/zshenv: system-wide .zshenv file for zsh(1).
<several lines of comments>
# Global Order: zshenv, zprofile, zshrc, zlogin

if [[ -z "$PATH" || "$PATH" == "/bin:/usr/bin" ]]
then
        export PATH="/usr/local/bin:/usr/bin:/bin:/usr/games"
fi

# Automatically added by Bedrock Linux
. /bedrock/run/zprofile
$ sudo strat ubuntu mv /etc/zsh/zshenv /etc/zsh/zshenv.bak
$ strat -r ubuntu zsh -c 'echo $PATH' #Correctly set
/bedrock/bin:/usr/local/sbin:/usr/local/bin:/opt/sbin:/opt/bin:/usr/sbin:/usr/bin:/sbin:/bin
```
Essentially, `/etc/zsh/zshenv` sourcing `/bedrock/run/zprofile` is resetting the `PATH`. Because `/etc/zsh/zshenv` is sourced unconditionally (according to the `zsh` man page), there appears to be no way around this other than removing that line from the file (which, since it is automatically added, cannot be done manually short of removing the file completely).

I don't see any new problems appearing as a result of the above fix, but maybe I just haven't done enough testing. :-)

**Comment by paradigm** [comment]
I can reproduce the issue.  Your explanation is good; what's happening all makes sense.

I only use my zsh as an interactive shell where I explicitly want Bedrock's cross-stratum functionality in place.  When I `strat -r` it's never zsh.  All my scripts are in a more portable shell flavor.  Thus, I've never run into this.  However, `strat -r <stratum> zsh` is a perfectly valid workflow which we should support.

Some background to help explain why Bedrock currently works as it does in this area:

- As you probably already figured out, Bedrock wants certain environment variables to be set up.  `$PATH` is the obvious one, but there's others as well, such as zsh's `$fpath`.
- If Bedrock creates a configuration file *before* their associated packages are installed, a package manager will (understandably) be upset upon install.  Instead, Bedrock injects its contents into various configuration when they're read.  This is why you cannot manually remove those lines - Bedrock will guess you (re)installed the package and re-inject the lines.
- Zsh does not necessarily read `/etc/profile`.  The expectation is for distros to configure zsh to do so, but some distros do not.  Thus Bedrock cannot simply place its configuration in that file.
- Some distros create `/etc/zsh/zshenv` on zsh install, but some do not.  Thus, Bedrock cannot rely on that for things to just work.  I found some which do not create `/etc/zsh/zshenv` do create `/etc/zsh/zprofile`.  [For example, see Arch's zsh package](https://www.archlinux.org/packages/extra/x86_64/zsh/files/).  Thus, Bedrock tries to inject into both, hoping at least one will exist.

Your work-around might work because of any of the following:

- Your zsh-providing distro packages configuration to read one of the other files Bedrock alters.
- Your interactive shell isn't zsh.
- Your running zsh instances all inherited Bedrock's env var stuff from before your change.  Once you reboot they'll be gone.

Your work-around may not be viable for you if it's that last situation.  If you haven't found it already, take a look at `/bedrock/etc/bedrock.conf`'s `[etc-inject]` section.  You can disable the automatic change here by commenting out or removing the `zsh/zshenv` line.  You can then manually add `. /bedrock/run/zprofile` to the appropriate zsh configuration file for your needs so your various variables are set up for your interactive shell.  However, for reasons described above, that's not a good general solution.

I currently see two ideas to fix this more generally:

- Have `/bedrock/run/zprofile` detect if it's a login shell.  If not, skip setting anything.  This should have it act more in line with how bash works in your example.  However, that won't make `strat -r ubuntu bash -l` or `strat -r ubuntu zsh -l` work.  Despite that limitation, this is where I'm currently leaning.  My guess is `strat -r` with login shells will be a very rare workflow, and one we can reasonably document as unsupported.
- Have `strat -r` set some inheritable property on the process that can be detected in `/bedrock/run/(z)profile` and used to skip changing environment variables.  For example, a new environment variable or a high numbered file descriptor.  However, there are valid workflows which will erase either of those.

I'll think about it and see if I can come up with another solution which does not have workflows that will break it.

**Comment by paradigm** [comment]
0.7.3 should include a fix.  If you can, please try to see if it resolves it for you and let me know.

**Comment by cptpcrd** [comment]
Yes, it appears to work. Thanks!
